### PR TITLE
Move exome tool functions (resolves #304)

### LINKED ISSUE #304: Move exome tool functions to tool library

### PR DESCRIPTION
**Pull request by jvivian**

### CODE REVIEW

**Comment by fnothaft** [comment]
I see many preprocessing functions here that seem to dupe the `gatk_germline` package. Is it correct to assume that the `gatk_germline` package is going to go away, or at least be restructured?

**Comment by jvivian** [comment]
> Is it correct to assume that the gatk_germline package is going to go away, or at least be restructured?

That is my understanding — @jpfeil will add / use job functions from the tool library and gatk_germline will be turn into a wrapper to handle inputs and perform wiring.

**Comment by fnothaft** [comment]
>> Is it correct to assume that the gatk_germline package is going to go away, or at least be restructured?
>
> That is my understanding — @jpfeil will add / use job functions from the tool library and gatk_germline will be turn into a wrapper to handle inputs and perform wiring.

Ah, to clarify: will those functions go away in this PR, or in a later PR (by @jpfeil, presumably), or...? I ask, because a cursory glance reveals that the GATK calls in this PR don't match the calls in the `gatk_preprocessing` pipeline (e.g., GATK 3.4 here vs [GATK 3.5 there](https://github.com/BD2KGenomics/toil-scripts/blob/master/src/toil_scripts/gatk_processing/gatk_preprocessing.py#L522)). I'd prefer to root out these differences now, or failing that, have a solid plan for rooting out the differences later.

**Comment by jvivian** [comment]
> GATK calls in this PR don't match the calls in the gatk_preprocessing pipeline (e.g., GATK 3.4 here vs GATK 3.5 there).

Ah yes, that is an issue. I can bump the GATK version to 3.5, but the commands being called should be the same unless there's some germline-specific parameters that GATK preprocessing requires. I see for example that RTC in **gatk_preprocessing** has the arg:

```
'-U', 'ALLOW_SEQ_DICT_INCOMPATIBILITY'
```

@jpfeil — You down to compare the functions I've moved to the common tool library to the ones in gatk_preprocessing?  We can figure out how best to tweak them to accommodate both the somatic and germline pipelines.

**Comment by fnothaft** [comment]
Perhaps it'd be best for you, me, and @jpfeil to sit down (on the phone, presumably) and to go through the differences in command invocations with a fine toothed comb.

**Comment by jvivian** [comment]
@hannes-ucsc — Ready for review

**Comment by fnothaft** [comment]
I'll make a review pass as well. I'll try to get it today, but is it OK if I make the pass tomorrow? I have to take my daughter to the Dr's today for checkup and vaccinations, so I'm about 25% online...

**Comment by jvivian** [comment]
@fnothaft — Definitely. I have several small PRs that shouldn't cause any conflicts that need merging, and ideally the bwa-refactor can be merged before this which will require me to re-rebase this PR.

**Comment by fnothaft** [comment]
SGTM! I'll try to make a pass today, but we'll see.

**Comment by hannes-ucsc** [comment]
Deferring to detailed review to @fnothaft.

**Comment by jvivian** [comment]
Rebased onto master

File: src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py [hunk 1]
```diff
@@ -22,7 +22,7 @@
 from toil_scripts.lib.urls import download_url_job, s3am_upload
 from toil_scripts.tools.QC import run_fastqc
 from toil_scripts.tools.aligners import run_star
-from toil_scripts.tools.preprocessing import cutadapt
+from toil_scripts.tools.preprocessing import run_cutadapt
```
Comment (fnothaft) [inline src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py]: Was this change supposed to get in to this PR? Seems unrelated...

File: src/toil_scripts/tools/__init__.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import os
+import subprocess
+
+
+def get_mean_insert_size(work_dir, bam_name):
+    """Function taken from MC3 Pipeline"""
+    cmd = "docker run --log-driver=none --rm -v {}:/data quay.io/ucsc_cgl/samtools " \
```
Comment (fnothaft) [inline src/toil_scripts/tools/__init__.py]: This is a thing you can do, and it's functionally correct, but can we open a ticket to replace this code with something that doesn't involve us munging samtools output? It just seems very fragile.

File: src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py (hunk 1 above)
Comment (jvivian) [inline src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py]: Just one of those things I noticed while working on this PR, felt too unimportant to move to its own PR. Happy to remove it if you want me to.

File: src/toil_scripts/tools/mutation_callers.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    for file_store_id, name in zip(file_ids, file_names):
+        job.fileStore.readGlobalFile(file_store_id, os.path.join(work_dir, name))
+    # Call: MuTect
+    parameters = ['--analysis_type', 'MuTect',
+                  '--reference_sequence', 'ref.fasta',
+                  '--cosmic', '/data/cosmic.vcf',
+                  '--dbsnp', '/data/dbsnp.vcf',
+                  '--input_file:normal', '/data/normal.bam',
+                  '--input_file:tumor', '/data/tumor.bam',
+                  '--tumor_lod', str(10),
```
Comment (fnothaft) [inline src/toil_scripts/tools/mutation_callers.py]: Document magic numbers for both `tumor_lod`s.

File: src/toil_scripts/tools/mutation_callers.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    """
+    work_dir = job.fileStore.getLocalTempDir()
+    file_ids = [normal_bam, normal_bai, tumor_bam, tumor_bai, ref, fai]
+    file_names = ['normal.bam', 'normal.bai', 'tumor.bam', 'tumor.bai', 'ref.fasta', 'ref.fasta.fai']
+    for file_store_id, name in zip(file_ids, file_names):
+        job.fileStore.readGlobalFile(file_store_id, os.path.join(work_dir, name))
+    # Create Pindel config
+    with open(os.path.join(work_dir, 'pindel-config.txt'), 'w') as f:
+        for bam in ['normal.bam', 'tumor.bam']:
+            f.write('/data/{} {} {}\n'.format(bam, get_mean_insert_size(work_dir, bam), os.path.splitext(bam)[0]))
```
Comment (fnothaft) [inline src/toil_scripts/tools/mutation_callers.py]: Is the value of `os.path.splitext(bam)[0]` the sample ID? Can you document?

File: src/toil_scripts/tools/preprocessing.py [hunk 5]
```diff
@@ -45,4 +45,258 @@ def cutadapt(job, r1_id, r2_id, fwd_3pr_adapter, rev_3pr_adapter):
...
+        job.fileStore.readGlobalFile(file_store_id, os.path.join(work_dir, name))
+    # Call: GATK -- IndelRealigner
+    parameters = ['-T', 'IndelRealigner',
+                  '-R', '/data/ref.fasta',
+                  '-I', '/data/sample.bam',
+                  '-known', '/data/phase.vcf',
+                  '-known', '/data/mills.vcf',
+                  '-targetIntervals', '/data/sample.intervals',
+                  '--downsampling_type', 'NONE',
+                  '-maxReads', str(720000),
```
Comment (fnothaft) [inline src/toil_scripts/tools/preprocessing.py]: Document magic numbers.

**Comment by fnothaft** [comment]
Other than a few small nits, LGTM! Let me know when the nits are addressed and I'll merge.

File: src/toil_scripts/tools/__init__.py (hunk 2 above)
Comment (jvivian) [inline src/toil_scripts/tools/__init__.py]: I agree, I don't like it at all. #347

File: src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py (hunk 1 above)
Comment (fnothaft) [inline src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py]: Ah, no big deal then. It doesn't offend me, I just wanted to make sure it wasn't accidentally in the wrong place.

**Comment by fnothaft** [comment]
This LGTM. Can you squash down?

**Comment by fnothaft** [comment]
Also, @hannes-ucsc, this is good to merge on your side (after @jvivian) squashes, correct?

**Comment by jvivian** [comment]
@fnothaft— Should I squash to one commit?

**Comment by fnothaft** [comment]
> @fnothaft— Should I squash to one commit?

That would be my preference.

**Comment by hannes-ucsc** [comment]
+1

I'm tempted to make up another rule—in addition to all the other ones that only exist in my head— requiring that moving code around is done in a separate commit. Changes to the moved code should be in a subsequent commit. That way the reviewers would be able to discern what actually changed in the moved code after it was moved. Right now I just see wads of red and green.

**Comment by fnothaft** [comment]
> I'm tempted to make up another rule—in addition to all the other ones that only exist in my head— requiring that moving code around is done in a separate commit. Changes to the moved code should be in a subsequent commit. That way the reviewers would be able to discern what actually changed in the moved code after it was moved. Right now I just see wads of red and green.

 I would +1 that rule.

**Comment by jvivian** [comment]
+1 to that as well

**Comment by fnothaft** [comment]
We should put the rules in a `CONTRIBUTING.md` file ;).

**Comment by hannes-ucsc** [comment]
> We should put the rules in a CONTRIBUTING.md file ;).

We track that for Toil as https://github.com/BD2KGenomics/toil/issues/849. The one for toil_script should be very similar.
